### PR TITLE
fix: preserve multiline docstring indentation under rules/checkpoints

### LINKED ISSUE #303: Multiline docstrings under rules and checkpoints are incorrectly deindented
Multiline top-level rule comments are deindented incorrectly (or at least differently than prior to v2 - I prefer the old way to keep indentation).

This:

```snakemake
rule test:
    """
    Comment
    """
    input:
        "test.txt",
```

gets formatted to this:

```snakemake
rule test:
    """
Comment
"""
    input:
        "test.txt",
```

This comment pattern is used a lot in Nextstrain workflows where it'd add a lot of churn.

This happens on both version v2.0.0 and v2.0.1

### PR DESCRIPTION
**Pull request by mbhall88**
closes #303

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of docstring indentation and alignment across Python and non-Python contexts, preserving relative indentation, blank lines, and comment spacing inside rules/checkpoints.

* **Tests**
  * Expanded test coverage for multiline and single-line docstrings, nested contexts, blank-line behavior, and edge cases to ensure consistent formatter output.

### CODE REVIEW

File: tests/test_formatter.py [hunk 1]
```diff
@@ -984,6 +978,59 @@ def test_docstrings_get_retabbed_for_snakecode_only(self):
 '''
         assert formatter.get_formatted() == expected
 
+    def test_multiline_docstring_inside_rule_preserves_indentation(self):
+        """https://github.com/snakemake/snakefmt/issues/303"""
+        snakecode = f"""rule test:
+{TAB * 1}""\"
```
Comment (Hocnonsense) [inline tests/test_formatter.py]: Suppose to convert new line to `\n` so the style of tests keeps?

File: snakefmt/formatter.py [hunk 2]
```diff
@@ -368,7 +373,9 @@ def run_black_format_str(
             fmted = textwrap.dedent(s)
         return fmted
 
-    def align_strings(self, string: str, target_indent: int) -> str:
+    def align_strings(
```
Comment (Hocnonsense) [inline snakefmt/formatter.py]: (may be out of scope) I notice that `self` is never used in `align_strings`, so it should be a normal function instead of a method of `Formatter`

File: tests/test_formatter.py [hunk 3]
```diff
@@ -984,6 +978,59 @@ def test_docstrings_get_retabbed_for_snakecode_only(self):
...
+{TAB * 1}Comment
+{TAB * 1}""\"
```
Comment (Hocnonsense) [inline tests/test_formatter.py]: (may be out of scope) I found another bug when testing this:
```python
        snakecode = f"""rule test:
{TAB * 1}""\"
{TAB * 1}Comment
{TAB * 1}""\" # comment
...
"""
```

**Review by Hocnonsense — CHANGES_REQUESTED** [review]
Thanks for the fix! The test case clearly reproduces the issue.

However, I don't think this catches the root cause of the bug. The string passed to `align_strings` is already incorrectly formatted: the indentation has been stripped (`self.buffer`=`'"""\n    Comment\n    """\n'` -> `formatted`=`'"""\nComment\n"""\n'`) by `run_black_format_str` called at: https://github.com/snakemake/snakefmt/blob/c981289c631730b6e9fa9c286c48accb356f5bab/snakefmt/formatter.py#L104-L112

In `run_black_format_str`, `artificial_nest` is set to False for non-Python contexts (when neither `self.from_python` nor `self.syntax.accepts_python_code` is `True`). As a result, `self.buffer` is passed to Black unprotected, so Black treats it as a module-level docstring and strips the indentation by mistake.

https://github.com/snakemake/snakefmt/blob/c981289c631730b6e9fa9c286c48accb356f5bab/snakefmt/formatter.py#L309-L317

I'd suggest fixing this at the source:
```python
        artificial_nest = (
            target_indent > 0
            and not is_all_comments(string)
            and not no_nesting
            and len(string.strip().splitlines()) > 1
        )
        if artificial_nest:
            if self.from_python or self.syntax.accepts_python_code:
                artificial_nest = True
            elif self.vocab:
                # Check if this is a rule/checkpoint docstring.
                # If so, wrap in `if x:` to prevent Black from
                # treating it as a module-level docstring and stripping indentation.
                not_docstring = docstring_matcher.sub("", string).strip()
                artificial_nest = (not not_docstring) or is_all_comments(not_docstring)
            else:
                artificial_nest = False
```

--------------

The fix proposed above cannot pass all tests. After further investigation, I think it is better to handle rule/checkpoint docstrings in `flush_buffer` before calling `run_black_format_str` at L105. Specifically, when `self.keyword_indent > 0`, `self.vocab` is set, and the buffer is a plain docstring, we can bypass Black entirely and apply indentation directly via a dedicated `_indent_docstring` helper. This keeps the docstring handling self-contained and avoids mixing it into `run_black_format_str` or `align_strings`.

**Comment by mbhall88** [comment]
Thanks for the thorough review, @Hocnonsense! Pushed a follow-up commit addressing the feedback.

### Root-cause fix
You were right that the original change patched a symptom. I confirmed your diagnosis by tracing the buffer: Black receives `'"""\n    Comment\n    """\n'` and returns `'"""\nComment\n"""\n'`, stripping the indent because it treats a rule/checkpoint docstring as a module-level docstring.

The fix now lives in `flush_buffer`: rule/checkpoint docstrings are detected and re-indented directly via a dedicated `indent_docstring` helper, and the `align_strings` change has been reverted.

One deliberate difference from the suggested approach: I kept Black for normalisation rather than bypassing it entirely. A full bypass breaks `test_docstrings_get_retabbed_for_snakecode_only`, which relies on Black collapsing `"""x\n"""` → `"""x"""` and normalising quotes — I believe this is the same wall your first attempt hit ("cannot pass all tests"). So the flow is now: let Black normalise (it flattens to column 0), then restore the keyword indent with `textwrap.indent`, which also leaves blank lines empty (addresses @coderabbitai's nitpick).

### Comment handling
While testing I found that a docstring followed by a comment was also broken (de-indented out of the rule), so `is_docstring` now also accepts a docstring followed only by comments. Covered by a new test.

### Tests
- Rewrote the docstring tests in the conventional per-line string style.
- Added unit tests for the new `is_docstring` / `indent_docstring` helpers.
- Added edge cases: docstring + comment, single-line collapse, nested rule, blank lines, relative indentation.

`make fmt` / `make lint` clean; full suite 344 passed, 1 xfailed.

### Out-of-scope items → follow-up issues
- The same-line-comment bug (`""" # comment`) is a pre-existing parser bug (the closing `"""` gets duplicated into `""""""` before Black sees it) — filed as #305.
- `align_strings` not using `self` → #306.

File: tests/test_formatter.py (hunk 3 above)
Comment (mbhall88) [reply tests/test_formatter.py]: have opened https://github.com/snakemake/snakefmt/issues/305 to address this

File: snakefmt/formatter.py (hunk 2 above)
Comment (mbhall88) [reply snakefmt/formatter.py]: have opened https://github.com/snakemake/snakefmt/issues/306 to address this

**Review by Hocnonsense — APPROVED** [review]
Clean and well-scoped fix, looks great to me!
